lib/Monad_WP/OptionMonad.thy

```diff
@@ -62,6 +62,30 @@ lemma opt_map_upd_Some:
 
 lemmas opt_map_upd[simp] = opt_map_upd_None opt_map_upd_Some
 
+lemma opt_map_Some_comp[simp]:
+  "f ||> h o g = f ||> g ||> h"
+  by (fastforce simp: opt_map_def split: option.split)
+
+lemma opt_map_fold_l:
+  "(id |> g) (f x) = (f |> g) x"
+  by (clarsimp simp: opt_map_def)
+
+(* LHS is the same as (f ||> id) *)
+lemma opt_map_unit_r[simp]:
+  "(f |> Some) x = f x"
+  by (clarsimp simp: opt_map_def split: option.split)
+
+lemma opt_map_unit_l[simp]:
+  "(Some |> f) x = f x"
+  by (clarsimp simp: opt_map_def split: option.split)
+
+lemma opt_map_zero_l[simp]:
+  "(Map.empty |> g) x = None"
+  by (clarsimp simp: opt_map_def)
+
+lemma opt_map_zero_r[simp]:
+  "(f |> Map.empty) x = None"
```

Comment:
And in the spirit of eta contraction for more applicability, this one would become `f |> Map.empty = Map.empty`, which shows that `empty` absorbs, which is also a nice algebraic property to have (and `Some` as the identity).

Reply:
If we don't have it yet, this suggest we might also want `Map.empty |> f = Map.empty` in the simp set.